packages/interface/src/utils/publishBatch.ts

```diff
@@ -0,0 +1,174 @@
+} from "maci-cli/sdk";
+import {
+  IG1ContractParams,
+  IMessageContractParams,
+  Keypair,
+  PCommand,
+  PrivKey,
+} from "maci-domainobjs";
+import { genRandomSalt } from "maci-crypto";
+import { publicClient } from "./permissionless";
+import { SmartAccountClient } from "permissionless";
+import { EntryPoint } from "permissionless/types";
+import { KernelEcdsaSmartAccount } from "permissionless/accounts";
+import { Address, HttpTransport, Chain } from "viem";
+// import { banner, validateSalt } from "node_modules/maci-cli/build/ts/utils"; // TODO: (merge-ok) this import breaks the frontend
+
+const MESSAGE_TREE_ARITY = 5;
+
+type ISmartAccountPublishBatchArgs = IPublishBatchArgs & {
+  smartAccount: KernelEcdsaSmartAccount<EntryPoint, HttpTransport, Chain>;
+  smartAccountClient: SmartAccountClient<EntryPoint, HttpTransport, Chain>;
+};
+
+
+/**
+ * @notice copied from maci-cli/sdk to add sponsorship
+ * Batch publish new messages to a MACI Poll contract
+ * @param {IPublishBatchArgs} args - The arguments for the publish command
+ * @returns {IPublishBatchData} The ephemeral private key used to encrypt the message, transaction hash
+ */
+export const publishBatch = async ({
+  messages,
+  pollId,
+  maciAddress,
+  publicKey,
+  privateKey,
+  signer,
+  quiet = true,
+  smartAccount,
+  smartAccountClient,
+}: ISmartAccountPublishBatchArgs): Promise<IPublishBatchData> => {
+  // TODO: (merge-ok) should this be added back?
+  // banner(quiet);
+
+  if (!PubKey.isValidSerializedPubKey(publicKey)) {
+    throw new Error("invalid MACI public key");
+  }
+
+  if (!PrivKey.isValidSerializedPrivKey(privateKey)) {
+    throw new Error("invalid MACI private key");
+  }
+
+  if (pollId < 0n) {
+    throw new Error(`invalid poll id ${pollId}`);
+  }
+
+  const userMaciPubKey = PubKey.deserialize(publicKey);
+  const userMaciPrivKey = PrivKey.deserialize(privateKey);
+  const maciContract = MACIFactory.connect(maciAddress, signer);
+  const pollContracts = await maciContract.getPoll(pollId);
+
+  const pollContract = PollFactory.connect(pollContracts.poll, signer);
+
+  const [treeDepths, coordinatorPubKeyResult] = await Promise.all([
+    pollContract.treeDepths(),
+    pollContract.coordinatorPubKey(),
+  ]);
+  const maxVoteOptions = Number(
+    BigInt(MESSAGE_TREE_ARITY) ** treeDepths.voteOptionTreeDepth
+  );
+
+  // validate the vote options index against the max leaf index on-chain
+  messages.forEach(({ stateIndex, voteOptionIndex, salt, nonce }) => {
+    if (voteOptionIndex < 0 || maxVoteOptions < voteOptionIndex) {
+      throw new Error("invalid vote option index");
+    }
+
+    // check < 1 cause index zero is a blank state leaf
+    if (stateIndex < 1) {
+      throw new Error("invalid state index");
+    }
+
+    if (nonce < 0) {
+      throw new Error("invalid nonce");
+    }
+
+    // TODO: (merge-ok) add validation back here
+    // if (salt && !validateSalt(salt)) {
+    //   throw new Error("invalid salt");
```

Comment:
we can remove this, that was in case salts are sent as input but in this case they are randomly generated